### ISSUE TITLE
META - Ecosystem terrain standardization

### ISSUE THREAD

**Issue by Ryanf55** [ROS-Aerial/community#65]
# Design Goals

* Use the same terrain data across all devices in the ecosystem
  * Autopilot (PX4, ArduPilot, Etc)
  * Companion Computer (grid_map_geo) or other computers running ROS 2
  * Ground Station (QGC, MAVProxy, Mission Planner)
  * Simulator (Gazebo, SITL)
  * Developer visualization tool (RVIZ, Foxglove)
  * Autonomy navigation/planning frameworks (NAV2)
* Support loading terrain from an online terrain server
  * ArduPilot SRTM
  * AirMap
* Reduce network usage by maintaining a local cache of terrain data
  * Only download tiles once from online terrain server to devices in field
  * Cache terrain on the companion computer and/or autopilot to reduce GCS-Vehicle network congestion
* Reduce the time, complexity, and difficulty to set up terrain data for terrain navigation
  * Example: given no terrain data is installed on any devices, and the system is started without GPS signal, when the GPS is enabled, then the autopilot, companion computer and GCS retrieve terrain and the simulator displays the terrain and the local origin of the simulation is to the home location. 
* Enable using drones at large latitudes beyond SRTM's 60 degree limit 
* Support simulation of similar scale flight as [this](https://youtu.be/2glZyBssp44?si=FWaBcCorXjRMrHrR)
# Current Issues
### Gazebo
* Gazebo can't consume data from online terrain servers - the [Gazebo Classic tutorial](https://classic.gazebosim.org/tutorials?tut=dem&cat=build_world) involves a lengthy process to manually generate terrain meshes
* Gazebo can't visualize SRTM terrain tiles without decimation due to performance issues
  * Issue for not supporting /vsicurl/vsizip: https://github.com/gazebosim/gz-common/issues/596
  * Proposed feature for /vsicurl/vsizip: https://github.com/gazebosim/gz-common/pull/597
* Gazebo assumes the user wants to load the entire dataset supplied, which breaks down with large VRT's (it currently segfaults)
### QGC
* QGC uses Airmap terrain which is tied to Auterion's hosting and vulnerable to outages
  * Ticket which includes my GDAL proposal: https://github.com/mavlink/qgroundcontrol/issues/7774
* Only 2D
* Unlear support for terrain in Norway: https://github.com/mavlink/qgroundcontrol/issues/9280
* Unclear support for terrain in Finland: https://github.com/mavlink/qgroundcontrol/issues/7405

### MAVProxy
* Only 2D
* Same problems with 60 degrees latitude: https://github.com/ArduPilot/MAVProxy/issues/709
* Current docs are quite good and explain the features: https://ardupilot.org/mavproxy/docs/modules/terrain.html#terrain-reporting
  * Docs don't show how to use a different terrain server - maybe it's not possible
### Mission Planner
* Unclear support for terrain in Norway: 
* https://github.com/ArduPilot/MissionPlanner/issues/2527
* https://github.com/ArduPilot/MissionPlanner/pull/3338
## Companion Computer
* If the companion computer doesn't have SIM card, it can't fetch terrain data from online, even if the GCS has terrain data
## Autopilot
* Autopilot's GPS sensors may use different datums
  * ArduPilot's Location (in 4.5) [does not specify altitude datums](https://github.com/ArduPilot/ardupilot/blob/60369b893a10a57204af61314c60426aa51a26f8/libraries/AP_Common/Location.h#L23)
  * Instead, you should use `get_undulation()` as recommended [here by Tridge](https://github.com/ArduPilot/ardupilot/pull/21075#issuecomment-1208825466)
  * ArduPilot used to not protect users against missing terrain tiles
    * https://github.com/ArduPilot/ardupilot/issues/12923
# What's working well
## Terrain transmission between autopilot and GCS
* Based on MAVLink terrain protocol

## Navigation/Planning (NAV2)

* NAV2 currently is heavily tied to Costmap2D in the GlobalPlanner
* NAV2 can't support 3D planning
* NAV2 assumes grid-based algorithms, no allowance for geometric
* We want to efficiently represent airspace and coastlines without a grid

# Current Terrain Data Transmission Approaches

These are the standards or techniques to share terrain data between devices over networks, both IP and serial, wired and wireless:
* [MAVLink terrain protocol](https://mavlink.io/en/services/terrain.html)
* [GDAL virtual network drivers](https://gdal.org/user/virtual_file_systems.html)
* Airmap protocol - TODO find link
* https://cesium.com/platform/cesium-ion/content/cesium-world-terrain/
# Current Terrain Sources
* [ArduPilot TerrainGen](https://terrain.ardupilot.org/)  - [source code](https://github.com/ArduPilot/terraingen)
* [Auterion](https://terrain-ce.suite.auterion.com/api/v1/carpet) used by [QGC](https://github.com/mavlink/qgroundcontrol/blob/eee119e0b363a314665e35c2a7daf3f0bd63f9ac/src/QtLocationPlugin/Providers/ElevationMapProvider.h#L54C50-L54C87)

**Comment by knmcguire** [comment]
Yeah you'll need to select the empty template. but I'm meaning to remove the community's template issues... no-one actually uses those.

**Comment by tridge** [comment]
some other issues:
- ublox use a course grained table for alt models, its significantly wrong in many places
- i am not certain that terrain.ardupilot.org is AMSL everywhere, though the ublox issue complicates testing
- we shift terrain db up/down based on home vs terrain height

**Comment by Ryanf55** [comment]
I've been working with some ArduPilot partners flying in steep terrain who have issues with terrain inaccuracies and inefficient or unsafe flight paths.
So far, I created a Jupyter notebook for comparing rangefinder data to arbitrary terrain sources given an ArduPilot bin log. 
https://github.com/ArduPilot/ardupilot/pull/30089

The results for the Canadian yukon show that the the COP30 data is significantly better than SRTM1, so I created an issue to add that dataset to the ardupilot terrain server.
https://github.com/ArduPilot/ardupilot/pull/30089

Adding a GDAL backend QGC is also back in progress so that QGC users can easily change their terrain source.
